### PR TITLE
Add `ParameterExpression.numeric` to cast to number

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This is largely a wrap-up of the `if ParameterExpression.is_real(): ...` logic present in a few places across the library, but a bit more efficient for Symengine because it can re-use cases where it needed to force an evaluation to produce accurate results.  Mostly, though, it simplifies a few call sites.

### Details and comments

Depends on #11107, because `.evalf(real=True)` is a symengine-specific extension, and when `TemplateOptimization` produces mismatched `ParameterExpression`s it breaks this handling.

The additional handling in `numeric` to reduce the impact of the spurious imaginary components makes this generally more reliable. I came across problems with global phases during binding in a PR built on top of this - at the moment, the type of the `global_phase` in `QuantumCircuit.assign_parameters` isn't validated in the same way, but in a unification PR I had it was, and some deeply nested (but completely valid) assignments were failing.

### CODE REVIEW

**Comment by jakelishman** [comment]
Now rebased over #11107 and other changes to `main` (notably #10902) that have merged since this was created.  Ready for review / merge.

File: test/python/circuit/test_parameters.py [hunk 1]
```diff
@@ -2005,6 +2005,32 @@ def test_bound_expression_is_real(self):
...
+        self.assertIsInstance(one_int.numeric(), int)
+        self.assertEqual(one_int.numeric(), 1)
+        one_float = (1.0 + a).assign(a, 0.0)
+        self.assertIsInstance(one_float.numeric(), float)
+        self.assertEqual(one_float.numeric(), 1.0)
+        one_imaginary = (1j + a).assign(a, 0.0)
+        self.assertIsInstance(one_imaginary.numeric(), complex)
+        self.assertEqual(one_imaginary.numeric(), 1j)
+
+        # This is one particular case where symengine 0.9.2 (and probably others) struggles when
```
Comment (1ucian0) [inline test/python/circuit/test_parameters.py:2021]: This corner case can be removed with https://github.com/Qiskit/qiskit/pull/11315 ?

File: qiskit/circuit/parameterexpression.py [hunk 2]
```diff
@@ -495,6 +495,52 @@ def is_real(self):
...
+        # We've come across several ways in which symengine's general-purpose evaluators
+        # introduce spurious imaginary components can get involved in the output.  The most
+        # reliable strategy "try it and see" while forcing real floating-point evaluation.
+        try:
+            real_expr = self._symbol_expr.evalf(real=True)
+        except RuntimeError:
+            # Symengine returns `complex` if any imaginary floating-point enters at all, even if
+            # the result is zero.  The best we can do is detect that and decay to a float.
+            out = complex(self._symbol_expr)
+            return out.real if out.imag == 0.0 else out
```
Comment (1ucian0) [inline qiskit/circuit/parameterexpression.py:541]: should this be a [math.isclose](https://docs.python.org/3/whatsnew/3.5.html#pep-485-a-function-for-testing-approximate-equality)? Zero-checking floats is always hard to me...

File: qiskit/circuit/parameterexpression.py [hunk 3]
```diff
@@ -495,6 +495,52 @@ def is_real(self):
...
+
+        Returns:
+            A Python number representing the object.
+
+        Raises:
+            TypeError: if there are unbound parameters.
+        """
+        if self._parameter_symbols:
+            raise TypeError(
+                f"expression with unbound parameters '{self.parameters}' is not numeric"
```
Comment (1ucian0) [inline qiskit/circuit/parameterexpression.py]:
```suggestion
                f"Expression with unbound parameters '{self.parameters}' is not numeric"
```

File: qiskit/transpiler/passes/basis/basis_translator.py [hunk 4]
```diff
@@ -345,26 +337,9 @@ def _replace_node(self, dag, node, instr_map):
...
-                            raise TranspilerError(
-                                f"Global phase: {new_phase} is complex which is invalid"
-                            ) from exc
-                try:
-                    new_phase = float(new_phase)
-                except TypeError:
-                    pass
+                    new_phase = new_phase.numeric()
+                    if isinstance(new_phase, complex):
+                        raise TranspilerError(f"global phase must be real, but got '{new_phase}'")
```
Comment (1ucian0) [inline qiskit/transpiler/passes/basis/basis_translator.py]:
```suggestion
                        raise TranspilerError(f"Global phase must be real, but got '{new_phase}'")
```

File: qiskit/circuit/quantumcircuit.py [hunk 5]
```diff
@@ -3130,9 +3122,9 @@ def map_calibration(qubits, parameters, schedule):
                 for to_bind in contained:
                     parameter = parameter.assign(to_bind, parameter_binds.mapping[to_bind])
                 if not parameter.parameters:
-                    parameter = (
-                        int(parameter) if parameter._symbol_expr.is_integer else float(parameter)
-                    )
+                    parameter = parameter.numeric()
+                    if isinstance(parameter, complex):
+                        raise TypeError(f"calibration cannot use complex number: '{parameter}'")
```
Comment (1ucian0) [inline qiskit/circuit/quantumcircuit.py]:
```suggestion
                        raise TypeError(f"Calibration cannot use complex number: '{parameter}'")
```

**Review by 1ucian0 — COMMENTED** [review]
Looks good to me! Just some comments here and there.

File: qiskit/circuit/parameterexpression.py [hunk 6]
```diff
@@ -495,6 +495,52 @@ def is_real(self):
...
+
+        In general, an :class:`int` is only returned if the expression only involved symbolic
+        integers.  If floating-point values were used during the evaluation, the return value will
+        be a :class:`float` regardless of whether the represented value is an integer.  This is
+        because floating-point values "infect" symbolic computations by their inexact nature, and
+        symbolic libraries will use inexact floating-point semantics not exact real-number semantics
+        when they are involved.  If you want to assert that all floating-point calculations *were*
+        carried out at infinite precision (i.e. :class:`float` could represent every intermediate
+        value exactly), you can use :meth:`float.is_integer` to check if the return float represents
+        and integer and cast it using :class:`int` if so.  This would be an unusual pattern;
```
Comment (ElePT) [inline qiskit/circuit/parameterexpression.py]:
```suggestion
        an integer and cast it using :class:`int` if so.  This would be an unusual pattern;
```

File: test/python/circuit/test_parameters.py (hunk 1 above)
Comment (jakelishman) [reply test/python/circuit/test_parameters.py:2021]: I'd leave it in regardless - it still works as a regression test for us.

File: qiskit/circuit/parameterexpression.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/circuit/parameterexpression.py:541]: Personally I'd say no - the only times we've seen uses where bad imaginary components enter is from bugs within Symengine itself (which this function works around), where the coefficient scales with the operation being done, so could be arbitrarily large. Auto clipping to zero also makes it tricky for people who were _expecting_ to deal with very small numbers (which floating point is good at).

As an interface thing, a user can always clip the output themselves if needed, so I think not clipping is the safer choice.

**Review by ElePT — APPROVED** [review]
LGTM (I'll add it to the queue in the interest of unblocking  #11428)
